---
title: "Class Activity -- Multiple Testing"
author: "Kim Siegmund"
date: "6/17/2020"
output: html_document
---

# {.tabset}

## 2-sample t-test: 10,000 genes 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if(!require("genefilter")) {BiocManager::install("genefilter")}
if(!require("qvalue")) {install.packages("qvalue")}
library(genefilter)
library(qvalue)
```

Simulation settings:    
* n1=n2=50 is the sample size in each group    
* In group 1, each observation has 10000 genes ~ N(mean_vec, sd_vec), where mean_vec and sd_vec are the vectors of 10000 means and standard deviations we computed from our data set.   
* In group 2, each observation has 10000 genes ~ N(mean_vec+0.11*truediff, sd_vec), where truediff is a vector of length 10,000 (1 = true differential expression, 0 = null gene). When truediff is 0, the feature is simulated from the same parameter settings as in group 1.

We will assume 2500 genes are differentially expressed, and without loss of generality, simulate those to be in rows 1-2500. The last 7500 rows will be for genes not differentially expressed (nullgenes). 

Let's grab the means and standard deviations from our class data for the simulation.
```{r ReadData}
load("data/JBC 2012/jbcdat.rda")
set.seed(22)
idx   <-sample(c(1:47231),10000)
avg_e <- rowMeans(jbcdat$E[idx,])
sd_e  <- matrixStats::rowSds(jbcdat$E[idx,])
```


```{r biol_truth}
ng <- 10000           # number of genes
signgenes <- 2500     # number of significant genes
nullgenes <- ng-signgenes
# truenull is 1 if the gene is null and 0 if significant (the truth)
truenull <- c(rep(0,signgenes),rep(1,nullgenes))
```

Simulate the gene expression data.
```{r sim_datmat}
set.seed(25)
# 100 samples = nsamp
nsamp <- 100
n1 <- nsamp/2   
gp  <- factor(rep(c("group1", "group2"), each=n1))
# I'm going to sample everyone from the same expression profile, and then
# shift the value by mu_1 - mu_2 = 0.11 after for the differentially expressed genes
sdat <- replicate(nsamp,rnorm(n = ng, mean = avg_e, sd = sd_e))
# assuming differential expressed genes (deg) occur in first "signgenes" rows
sdat[c(1:signgenes),c(1:n1)] <- 
            sdat[c(1:signgenes),c(1:n1)] + 0.11
```

Comment: Using 0.11 to represent the difference in means for all significant genes is not meant to be realistic, but it allows us to practice our multiple test correction methods. We've also made the simplifying assumption that all features are independent. This is approximately true for gene expression studies since the correlated subsets are small relative to the size of the testing problem, but this may not apply in different contexts. Then alternative methods should be considered.

For each gene, perform a 2-sample t-test. 

```{r multttests}
tt <- genefilter::rowttests(sdat,gp)
```

```{r histogram}
hist(tt$p.value)
```

## 1. unadjusted p<0.05


```{r resultsrawp}
table(truenull,sign=tt$p.value < 0.05)
```

a. How many false-positives are called using a p<0.05 cutoff if there is no correction for multiple testing?


b. What is the false-discovery rate?



## 2. Bonferroni adjusted p<0.05

```{r bonferroni_correction}
Bonf.adjp <- p.adjust(tt$p.value,method="bonferroni")
table(truenull,sign=Bonf.adjp<0.05)
```

a. How many false-positives are there after a Bonferroni correction (adjusted p<0.05)?


b. What is the false-discovery rate?


c. How many differentially expressed genes do we find?


d. What is the power in this study?


## 3. Benjamini-Hochberg FDR adjusted p<0.05

```{r fdr_correction}
FDR.adjp <- p.adjust(tt$p.value,method="BH")
table(truenull,sign=FDR.adjp<0.05)
```


a. How many false-positives are there after a Benjamini and Hochberg FDR correction (FDR adjusted p<0.05)?


b. What is the false-discovery rate?


c. How many differentially expressed genes do we find?


d. What is the power in this study?


## 4. q < 0.05

```{r qv}
qv <- qvalue(tt$p.value)$qvalue
table(truenull,sign=qv<0.05)
```

a. How many false-positives are there (q<0.05)?


b. What is the false-discovery rate?


c. How many differentially expressed genes do we find?


d. What is the power in this study?


## Session Information
```{r sI}
sessionInfo()
```
