---
title: "Dimension Reduction Techniques"
author: "K Siegmund"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# {.tabset}

## R Libraries

```{r rlibraries}
if (!require(ggplot2)) install.packages("ggplot2")
if (!require(matrixStats)) install.packages("matrixStats")
if (!require(plotly)) install.packages("plotly")
if (!require(Rtsne)) {install.packages("Rtsne")}
if (!require(devtools)) install.packages("devtools")
if (!require(gg3D)) devtools::install_github("AckerDWM/gg3D")
if (!require(uwot)) install.packages("uwot")
if (!require(gridExtra)) install.packages("gridExtra")
library(limma)
library(ggplot2)
library(matrixStats)
library(plotly)
library(Rtsne)
library(devtools)
library(gg3D)
library(uwot)
library(gridExtra)
```


```{r loaddata}
load("../data/JBC 2012/jbcdat.rda")
```

## 2-D MDS 

The function plotMDS was written specifically for gene expression data so it expects the data to be formatted with genes in the rows and samples in the columns. It also has a default variable selection filter that we'll discuss in more detail later. The default is to select 500 features.
```{r MDSplot,fig.cap='MDS plot of 24 gene expression arrays using 500 most variable features.'}
limma::plotMDS(jbcdat$E,pch=16,
        col=unclass(jbcdat$targets$type),
        xlim = c(-1.5,1.5), ylim=c(-1,1),
        main="MDS plot") #color by type
legend(-1.5,1,levels(jbcdat$targets$type),
        pch=16,col=order(levels(jbcdat$targets$type)),
       cex=.75)
```

## 2-D PCA

The PCA analysis expects data formatted in the traditional format, with samples in the rows and features in the columns. 
```{r pca}
# transpose the expression matrix
tE <- t(jbcdat$E)
dim(tE)
```

Scaling of the data is important for PCA. A PCA of the covariance matrix is different than a PCA of the correlation matrix (using standardized variables). Gene expression data can have outliers so we will use a robust measure to scale the data. I'll use median absolute deviation (mad).

```{r sdzpca}
sdztE <- scale(tE,center=TRUE,scale=
                 matrixStats::colMads(tE))
```

In our summary variables, let's try to capture the biological variation between treatment groups when we reduce the dimension. If successful, our treatment groups might be visible on just a few variables. Differential expression between treatment groups should increase the variability of the expression values when measured across all samples. Let's rank genes on variation and pick the 500 most variable genes for the PCA. I will rank them based on median absolute deviation, which is less susceptible to outliers than ranking on variance.

```{r top500}
fmad  <- matrixStats::rowMads(jbcdat$E)
rfmad <- rank(-fmad)
fidx <- which(rfmad <= 500)
```

```{r top500pca,fig.cap='PCA of 24 gene expression arrays using 500 most variable features.'}
my.pca <- prcomp(sdztE[,fidx],retx=TRUE)
dfx <- as.data.frame(my.pca$x)

ggplot(dfx,  aes(x=PC1, y=PC2, color = jbcdat$targets$type )) + geom_point(size=2.5) +
  labs(color="Treatment")
```

We do see the samples from the different treatment groups cluster separately. Would we still see this if we use all 47k+ genes?

```{r pcaall, fig.cap='PCA of 24 gene expression arrays using 47,231 features.'}
my.pca <- prcomp(sdztE,retx=TRUE)
dfx <- as.data.frame(my.pca$x)

ggplot(dfx,  aes(x=PC1, y=PC2, color = jbcdat$targets$type )) + geom_point(size=2.5) + labs(color="Treatment")
```

No, we don't. Why not?  (We'll come back to this in 2 weeks.)


## 3-D PCA 

Let's see if the PCA plot can separate the overlapping groups if we consider more dimensions.

```{r 3dggplot, echo=FALSE, fig.cap='3-Dimensional PCA of 24 gene expression arrays (500 most variable features).'}
my.pca <- prcomp(sdztE[,fidx],retx=TRUE)
dfx <- as.data.frame(my.pca$x)

ggplot(dfx, aes(x=PC1, y=PC2, z=PC3, color=jbcdat$targets$type)) + theme_void() +
  axes_3D() +
  stat_3D(size=2.5) +
  labs(color="Treatment")
```

It does!   And I'm including a figure using the plotly library. Check out what it can do.

```{r 3dplotly}
plot_ly(x=dfx$PC1, y=dfx$PC2, z=dfx$PC3, type="scatter3d", mode="markers", color=jbcdat$targets$type)
```

## tSNE

tSNE is a more flexible method to visualize neighboring observations in lower dimensional space. Can we display the 6 different treatment groups in just 2 dimensions?

There is more than one package for running this. I'm going to run the 'Barnes-Hut implemented t-SNE'. This becomes very slow when n>>10^5. (Not a problem for our n=24!)

```{r tsne}
# Set a seed if you want reproducible results
set.seed(48)
tsne_out <- Rtsne::Rtsne(sdztE[,fidx],perplexity=7,theta=0.0) 
dfy <- as.data.frame(tsne_out$Y)
```

Here, I reduced the perplexity from the default (30) to 7. A rule of thumb about perplexity is that it should always be less than the sample size. For this data we actually know there are 4 replicates per group. theta sets the speed/accuracy trade-off. For exact TSNE theta=0.

```{r tsneplot, echo=FALSE, fig.cap='tSNE of 24 gene expression arrays - perplexity = 7, top 500 variable genes.'}
# Show the objects in the 2D tsne representation

ggplot(dfy,  aes(x=V1, y=V2, color = jbcdat$targets$type )) + geom_point(size=2.5) +
  theme(text = element_text(size=14)) +
   labs(color="Treatment")
```

What is different about this figure compared to the PCA?

And, let's check the robustness of this result using a different starting seed for the random number generator.
```{r tsne2}
# Set a seed if you want reproducible results
set.seed(45)
tsne_out <- Rtsne::Rtsne(sdztE[,fidx],perplexity=7,theta=0.0) 
dfy <- as.data.frame(tsne_out$Y)
```


```{r tsneplot2, echo=FALSE, fig.cap='2nd tSNE of 24 gene expression arrays - perplexity = 7, top 500 variable genes.'}
# Show the objects in the 2D tsne representation
ggplot(dfy,  aes(x=V1, y=V2, color = jbcdat$targets$type )) + geom_point(size=2.5) +
  theme(text = element_text(size=14)) +
  labs(color="Treatment")
```

It looks similar.  There are some default settings 'under the hood' that make this more stable. Note: pca = TRUE reduces the 500 features to just initial_dims = 50.

Let's stop and play with the settings. What if we set a smaller perplexity to look for finer structure?  Recall that perplexity controls the neighborhood size.


```{r tsne-perplex}
set.seed(45)
tsne_out1 <- Rtsne::Rtsne(sdztE[,fidx],perplexity=4,theta=0.0) 
tsne_out1 <- cbind.data.frame(tsne_out1$Y, perplex="Perplexity = 4")
set.seed(45)
tsne_out2 <- Rtsne::Rtsne(sdztE[,fidx],perplexity=3,theta=0.0) 
tsne_out2 <- cbind.data.frame(tsne_out2$Y, perplex="Perplexity = 3")

dfy <- rbind.data.frame(rbind(tsne_out1,tsne_out2))
colnames(dfy) <- c("V1","V2","perplex")

ggplot(dfy, aes(x=V1, y=V2, 
      color = c(jbcdat$targets$type,jbcdat$targets$type) ) ) + 
        geom_point() + 
        facet_wrap(~ perplex, ncol = 2, scales = "free") +
        theme(legend.position = "bottom") +
        theme(text = element_text(size=14)) +
        labs(color="Treatment") +
        ggtitle("tSNE")
```

Yes, we see more local structure with lower perplexity. We should run these again with new seeds to see how stable the relative locations of the clusters are.   tsne is known for being good at local structure, but not always good at global structure. Still, this paper (https://pubmed.ncbi.nlm.nih.gov/31780648/) has suggestions for how to select parameter settings that lead to reproducible results. We're doing 3 of them here, by filtering the genes and using pca in 50 dimensions for initialization.  How do we know that?  Type ?Rtsne to see the default settings.

## UMAP

Now let's try UMAP.  There are different packages for running UMAP (libraries: UMAP and uwot). Here's the one that I think gets called by the single-cell RNAseq function, so it's the one we'll try.  See the github page for comments about reproducibility (and how many factors will affect it!).  https://github.com/jlmelville/uwot

Since R calls functions outside of R, the computations will also depend on the C++ compiler.

Let's set the number of neighbors to the perplexity from tSNE.

```{r umap}
# Show the objects in the 2D umap representation
set.seed(4987)
umap_out <- uwot::umap(sdztE[,fidx],n_neighbors=7,init="spca") 
```


```{r umapplot, echo=FALSE, fig.cap='UMAP of 24 gene expression arrays (using top 500 variable genes).'}
dfy <- as.data.frame(umap_out)
ggplot(dfy,  aes(x=V1, y=V2, color = jbcdat$targets$type )) + geom_point(size=2.5) +
  theme(text = element_text(size=14)) +
  labs(color="Treatment") +
  ggtitle("UMAP")
```

How does this figure compared to the PCA and tSNE plots?

And, let's check the robustness of this result using a different starting seed for the random number generator.
```{r umap2}
set.seed(6548)
umap_out <- uwot::umap(sdztE[,fidx],n_neighbors=7,init="spca") 
dfy <- as.data.frame(umap_out)
ggplot(dfy,  aes(x=V1, y=V2, color = jbcdat$targets$type )) + geom_point(size=2.5) +
  theme(text = element_text(size=14)) +
  labs(color="Treatment") +
  ggtitle("UMAP")
```

There is a different relative spacing of the 3 main clusters.

And like with tSNE, let's see what happens when we change neighborhood size.

```{r umap3}
set.seed(6548)
umap_out1 <- uwot::umap(sdztE[,fidx],n_neighbors=3,init="spca") 
umap_out1 <- cbind.data.frame(umap_out1, nneighbors="n_neighbors = 3")
set.seed(6548)
umap_out2 <- uwot::umap(sdztE[,fidx],n_neighbors=7,init="spca") 
umap_out2 <- cbind.data.frame(umap_out2, nneighbors="n_neighbors = 7")

dfy <- rbind.data.frame(rbind(umap_out1,umap_out2))
colnames(dfy) <- c("V1","V2","nneighbors")

ggplot(dfy, aes(x=V1, y=V2, 
      color = c(jbcdat$targets$type,jbcdat$targets$type) ) ) + 
        geom_point() + 
        facet_wrap(~ nneighbors, ncol = 2, scales = "free") +
        theme(legend.position = "bottom") +
        labs(color="Treatment") +
        ggtitle("UMAP")
```

Smaller neighborhoods finds more structure.

How stable are the results for a neighborhood size of 3 using two different random number seeds?

```{r umap4}
set.seed(6548)
umap_out1 <- uwot::umap(sdztE[,fidx],n_neighbors=3,init="spca") 
umap_out1 <- cbind.data.frame(umap_out1, seed="6548")
set.seed(20)
umap_out2 <- uwot::umap(sdztE[,fidx],n_neighbors=3,init="spca") 
umap_out2 <- cbind.data.frame(umap_out2, seed="20")

dfy <- rbind.data.frame(rbind(umap_out1,umap_out2))
colnames(dfy) <- c("V1","V2","seed")

ggplot(dfy, aes(x=V1, y=V2, 
            color = c(jbcdat$targets$type,jbcdat$targets$type) ) ) + 
        geom_point() + 
        facet_wrap(~ seed, ncol = 2, scales = "free") +
        theme(legend.position = "bottom") +
        labs(color="Treatment") +
        ggtitle("UMAP")
```

The groups show different relative positions, telling us that the relative position of the groups is NOT interpretable.

## sessionInfo

```{r sessionInfo}
sessionInfo()
```

