---
title: "Differential Expression - 2 groups"
author: "ks"
date: "`r Sys.Date()`"
output: html_document
---

# {.tabset}

## Data

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
if (!require(genefilter)) BiocManager::install("genefilter")
library(genefilter)
library(limma)
library(ggplot2)
```

Let's analyze the prostate cancer cell line data we've used in class. The treatments are:  

treatment 1: siNS, sip300, siCBP 

treatment 2:  0 hr,  16 hrs
     
All cells are treated with androgen at 0 hrs. The genes that change expression between 0 and 16 hours are genes that are regulated by androgen. These might vary based on the different treatments siNS (control), sip300, and siCBP. 

Let's compare average gene expression between 0 and 16 hrs for the control reaction.

Load the data (from JBC (2012)).
```{r ReadData}
jbcdir=c("../data/JBC 2012")
load(file.path(jbcdir,"jbcdat.rda"))
```

Here is how we'll subset the data to two time points from 1 treatment level (siNS).

```{r subset-data}
geoid <- rownames(jbcdat$targets)[jbcdat$targets$treatment == "siNS"]
subsetEobj <- function(dataobj,geonames){
  subobj <-NULL
     subobj$targets <- dataobj$targets[geonames,]
     subobj$E       <- dataobj$E[,geonames]
     subobj$genes   <- dataobj$genes
     subobj
}

siNSobj <- subsetEobj(jbcdat,geoid)
# check the annotation and expression are both subset identically
identical(rownames(siNSobj$targets),colnames(siNSobj$E))
```

## MA plot

The MA plot is a plot of the difference in average log2 expression  in group 1 vs average log2 expression in group 2 (a.k.a Log2 Fold Change) vs the overall average log2 expression.

```{r MAplot}
Index1 <- which(siNSobj$target$hour=="0hr")
Index2 <- which(siNSobj$target$hour=="16hr")

d <- rowMeans(siNSobj$E[,Index2])-rowMeans(siNSobj$E[,Index1])
a <- rowMeans(siNSobj$E)

par(mfrow=c(1,2))
plot(a,d,xlab="Average (log2) intensity",ylab="Log2 Fold Change",main="MA plot",pch=".")
smoothScatter(a,d,nrpoints=500,
              xlab="Average (log2) intensity",
              ylab="Log2 Fold Change",
              main="MA plot")
abline(-1,0,col=7,lwd=3)
abline(1,0,col=7,lwd=3)
abline(0,0,col=2,lwd=3)
```

The figure on the right is faster to print, and will be a smaller size file to save than the figure on the left.

## 2-sample t-tests

Here we do a single t-test for each feature in our data set.

```{r ttests}
time2 <- ifelse(siNSobj$targets$hour=="16hr",1,2)
tt <- genefilter::rowttests(siNSobj$E,factor(time2))
head(tt,n=3)
```

These are equivalent to the Ftest with equal variances: 

```{r Ftests}
Ft <- genefilter::rowFtests(siNSobj$E,factor(siNSobj$target$hour),var.equal=TRUE)
head(Ft,n=3)
```


## Volcano plot

The volcano plot is a scatterdiagram of -log10(P-value) vs log2 fold change.

```{r volcano-plot}
tt   <- genefilter::rowttests(siNSobj$E,factor(time2))
d    <- tt$dm
lodt <- (-log10(tt$p.value))

plot(d,lodt,main="Volcano plot for t-test",xlab="Log2 Fold Change",
	ylab="-log10 (P-value)",pch=".")
```


We can give emphasis to highly sigificant pvalues (-log10p > 5).

```{r volcanoPlot-highlightpts}
plot(d,lodt,main="Volcano plot for t-test",xlab="Log2 Fold Change",
	ylab="-log10 (P-value)",pch=".")
# make larger data points when -log10 pvalue > 5 
#  pch = 20 gives solid dots, col=4 is blue
points(d[lodt>5],lodt[lodt>5],pch=20,col=4)
```

And again, smoothed scatterplot can be used.  We can also highlight the top 10 features with the smallest p-values.

```{r smoothedVolcano}
smoothScatter(d,lodt,nrpoints=500,
              xlab="Log2 Fold Change",
              ylab="-log10 (P-value)",
              main="Volcano plot for t-test")
# If you want to highlight high lodt & fold change
# col= 6 is pink
rankp <- rank(tt$p.value)
text(d[rankp<=10],lodt[rankp<=10],
     siNSobj$genes$Symbol[rankp<=10],col="blue",cex=.7)
```

## 2-sample moderated t-tests 

Now let's see how the results change if we run a moderated t-test.
```{r modT}
design <- model.matrix(~factor(siNSobj$targets$hour))

fit  <- limma::lmFit(siNSobj$E,design)
efit <- limma::eBayes(fit)
```

The file 'differential expression supplemental code' will show the calculations for the moderated t-test.

```{r modTvolcanoplot, echo=FALSE}
par(mfrow=c(1,2))
lodmt <- -log10(efit$p.value[,2])
smoothScatter(efit$coef[,2],
              lodt,ylim=c(0,10.5),
              nrpoints=500,
              main="Volcano plot for t-test",xlab="Log2 Fold Change",
	ylab="-log10 (P-value)")


smoothScatter(efit$coef[,2],
              lodmt, ylim=c(0,10.5),
              nrpoints=500,
              main="Volcano plot for moderated t-test",xlab="Log2 Fold Change",
	ylab="-log10 (P-value)")
```

Name 2 differences you see comparing these 2 figures.

The limma package has a function for creating volcano plots that includes an option for labeling the top hits.

```{r volcanoplot}
volcanoplot(efit, coef = 2, style = "p-value", highlight = 10, names = siNSobj$genes$Symbol, hl.col="blue",
            xlab = "Log2 Fold Change", ylab = NULL, pch=16, cex=0.35)
```

## Compare t-test vs mod t

Let's compare the pvalues from student's t-test and the moderated t-test.
```{r ttvmodt}
par(mar=c(5,5,3,2))
plot(lodt,lodmt,pch=".",cex.axis=1.2,cex.lab=1.2,
	xlab="-log10(p), 2-sample t",
	ylab="-log10(p), moderated t")
abline(0,1,col=2,lwd=3)
```

That's a pretty noisy association.

Was there a change in the difference in group means (Log2 Fold change)?

Let's compare the moderated standard deviation to the sample sd.
```{r sdplot}
plot(efit$sigma,sqrt(efit$s2.post),pch=".",
     xlab="Pooled sd",ylab="Moderated pooled sd")
abline(0,1)
segments(sqrt(efit$s2.prior),0,sqrt(efit$s2.prior),1,col=2)
text(0.2,0.8,"sd Prior",col=2,adj=0,lwd=3)
```

This shows how the moderated (pooled) standard deviation is "shrunk" closer to the prior, $s_0$. So, small standard deviations become a little bigger, making the moderated t-test smaller and less statistically significant. Large standard deviations are also shrunk towards the prior. This makes large standard deviations smaller, which makes the moderated t-test larger and more statistically significant.

Let's look at the expression of a gene that was highly significant with the t-test and is no longer so when using the moderated t-test.

```{r selectgene}
thisone <- which(lodt > 4 & lodmt < 1.05)
index <- order(siNSobj$targets$hour)
data <- cbind.data.frame(Index=1:8,
                         log2E = siNSobj$E[thisone,index],
                         time = siNSobj$targets$hour[index])
ggplot(data,  aes(x=Index, y=log2E, color = time)) + geom_point(size=2.5) +
  labs(color="Time") + ylab("Log2 Expression")
```

Although the expression level does separate the samples by time, this is a very lowly expressed gene, and the difference in group means is very small.  It could be that this is a chance occurrence due to the small sample size.

Now let's look at the expression signal coming from a feature that was highly significant by both the t-test and moderated t-test.

```{r selectgene2}
thisone <- which(lodt > 7 & lodmt > 10)
index <- order(siNSobj$targets$hour)
data <- cbind.data.frame(Index=1:8,
                         log2E = siNSobj$E[thisone,index],
                         time = siNSobj$targets$hour[index])
ggplot(data,  aes(x=Index, y=log2E, color = time)) + geom_point(size=2.5) +
  labs(color="Time") + ylab("Log2 Expression")
```

Now that's a highly expressed gene, with a large difference in mean expression between the two groups.

## Results Table

The command topTable will summarize the regression results, ranked by (decreasing) p-value.

```{r TopT}
topT <- topTable(efit,coef=2,n=10)
topT
```

What information is provided on these features?

```{r genenames}
siNSobj$genes[rownames(topT),]
```

Notice that 2 of the top 10 hits are for the same Entrez Gene ID. And they have different associated gene Symbol (that's problematic!) 

## SessionInfo

```{r sessionInfo}
sessionInfo()
```